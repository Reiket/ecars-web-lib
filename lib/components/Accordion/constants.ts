export const ACCORDION_TEST_ID = 'accordionTestId';
