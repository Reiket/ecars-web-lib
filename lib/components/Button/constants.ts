import type {ButtonHTMLAttributes} from 'react';

export const BUTTON_SIZE = Object.freeze({
  BIG: 'big',
  SMALL: 'small',
});

export const BUTTON_COLOR = Object.freeze({
  GREEN: 'green',
  WHITE: 'white',
  GRAY: 'gray',
});


export type ButtonAttributesType = ButtonHTMLAttributes<HTMLButtonElement>['type']
export type ButtonSizeType = (typeof BUTTON_SIZE)[keyof typeof BUTTON_SIZE];
export type ButtonColorType = (typeof BUTTON_COLOR)[keyof typeof BUTTON_COLOR];

export const BUTTON_TEST_ID = 'buttonTestId';