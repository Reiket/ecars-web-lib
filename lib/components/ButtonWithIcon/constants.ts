export const LEFT_ICON_BUTTON_TEST_ID = 'leftIconButtonTestId';
export const RIGHT_ICON_BUTTON_TEST_ID = 'rightIconButtonTestId';
