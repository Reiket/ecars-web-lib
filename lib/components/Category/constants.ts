export const CATEGORIES_LIST = Object.freeze({
  GUIDES: 'guides',
  REVIEWS: 'reviews',
  NEWS: 'news',
});

export type CategoriesListType = (typeof CATEGORIES_LIST)[keyof typeof CATEGORIES_LIST];

export const CATEGORY_TEST_ID = 'categoryTestId';
