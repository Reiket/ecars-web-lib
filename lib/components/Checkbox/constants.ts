import type {ElementProps} from '@/services/types';

export interface CheckboxProps extends ElementProps {
  name?: string;
  id?: string;
  disabled?: boolean;
  onChange?: (checked: boolean) => void;
  checked?: boolean;
  hasError?: boolean;
}

export const CHECKBOX_TEST_ID = 'checkboxTestId';
export const CHECKBOX_TEST_INPUT_ID = 'checkboxTestInputId';
