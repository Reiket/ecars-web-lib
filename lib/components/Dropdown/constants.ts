import type {ReactNode} from 'react';

export interface DropdownOption {
  value: string;
  label?: ReactNode;
}

export interface DropdownProps {
  onSelect: (value: string) => void;
  category: string;
  options: DropdownOption[];
}

export const DROPDOWN_TEST_ID = 'dropdownTestId';

export const DROPDOWN_THEME = Object.freeze({
  DARK: 'dark',
  LIGHT: 'light',
});

export type DropdownThemeType = (typeof DROPDOWN_THEME)[keyof typeof DROPDOWN_THEME];
