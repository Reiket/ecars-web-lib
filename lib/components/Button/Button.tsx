import type {FC, MouseEvent, ReactNode} from 'react';
import {BUTTON_TEST_ID, ButtonAttributesType, ButtonColorType, ButtonSizeType} from '@/components/Button/constants';
import {cn} from '@/services/helpers';

export interface ButtonProps {
  size: ButtonSizeType;
  children: ReactNode;
  color: ButtonColorType;
  type?: ButtonAttributesType;
  disabled?: boolean;
  onClick?: (event: MouseEvent<HTMLButtonElement>) => void;
  isTransparent?: boolean;
  withIcon?: boolean;
}

export const Button: FC<ButtonProps> = ({
  children,
  size,
  color,
  onClick,
  type = 'button',
  disabled = false,
  isTransparent = false,
  withIcon = false,
}) => {
  const classNames = cn('button', `button-${size}`, `button-${color}`, {
    'button-transparent': isTransparent,
    'button-icon': withIcon,
  });
  return (
    <button
      data-testid={BUTTON_TEST_ID}
      className={classNames}
      disabled={disabled}
      type={type}
      onClick={onClick}
    >
      {children}
    </button>
  );
};
