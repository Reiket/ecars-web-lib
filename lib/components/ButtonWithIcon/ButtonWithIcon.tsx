import type {ComponentType, FC, ReactNode} from 'react';
import type {ButtonProps} from '@/components/Button/Button';
import {Button} from '@/components/Button/Button';
import {LEFT_ICON_BUTTON_TEST_ID, RIGHT_ICON_BUTTON_TEST_ID} from '@/components/ButtonWithIcon/constants';

interface IconProps {
  className?: string;
}

export interface ButtonIconProps<T> extends ButtonProps {
  children: ReactNode;
  RightIcon?: ComponentType<T>;
  LeftIcon?: ComponentType<T>;
  className?: string;
}

export const ButtonWithIcon: FC<ButtonIconProps<IconProps>> = ({
  RightIcon,
  LeftIcon,
  children,
  className,
  ...props
}) => (
  <Button
    withIcon
    {...props}
  >
    {!!LeftIcon && (
      <LeftIcon
        className={className}
        data-testid={LEFT_ICON_BUTTON_TEST_ID}
      />
    )}
    {children}
    {!!RightIcon && (
      <RightIcon
        className={className}
        data-testid={RIGHT_ICON_BUTTON_TEST_ID}
      />
    )}
  </Button>
);
