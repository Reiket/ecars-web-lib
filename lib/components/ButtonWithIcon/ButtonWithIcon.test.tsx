import {cleanup, render, screen} from '@testing-library/react';
import {BUTTON_COLOR, BUTTON_SIZE} from '@/components/Button/constants';
import {ButtonWithIcon} from '@/components/ButtonWithIcon/ButtonWithIcon';
import {Icons} from '@/services/icons';
import '@testing-library/jest-dom';
import {LEFT_ICON_BUTTON_TEST_ID, RIGHT_ICON_BUTTON_TEST_ID} from '@/components/ButtonWithIcon/constants';

describe('ButtonWithIcon component', () => {
  afterEach(cleanup);
  const props = {
    color: BUTTON_COLOR.GRAY,
    size: BUTTON_SIZE.BIG,
  };
  test('renders component correctly', () => {
    const {container} = render(<ButtonWithIcon {...props}>Click me</ButtonWithIcon>);
    expect(screen.getByText('Click me')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  test('renders LeftIcon when provided', () => {
    const {container} = render(
      <ButtonWithIcon
        LeftIcon={Icons.ArrowNarrowLeft}
        {...props}
      >
        Click
      </ButtonWithIcon>,
    );
    expect(screen.getByTestId(LEFT_ICON_BUTTON_TEST_ID)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('renders RightIcon when provided', () => {
    const {container} = render(
      <ButtonWithIcon
        RightIcon={Icons.ArrowNarrowRight}
        {...props}
      >
        Click
      </ButtonWithIcon>,
    );
    expect(screen.getByTestId(RIGHT_ICON_BUTTON_TEST_ID)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('renders both LeftIcon and RightIcon when provided', () => {
    const {container} = render(
      <ButtonWithIcon
        LeftIcon={Icons.ArrowNarrowLeft}
        RightIcon={Icons.ArrowNarrowRight}
        {...props}
      >
        Click
      </ButtonWithIcon>,
    );
    expect(screen.getByTestId(LEFT_ICON_BUTTON_TEST_ID)).toBeInTheDocument();
    expect(screen.getByTestId(RIGHT_ICON_BUTTON_TEST_ID)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
