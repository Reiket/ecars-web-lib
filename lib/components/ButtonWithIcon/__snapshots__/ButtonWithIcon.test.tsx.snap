// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonWithIcon component renders LeftIcon when provided 1`] = `
<div>
  <button
    class="button button--big button--gray button--icon"
    data-testid="buttonTestId"
    type="button"
  >
    <svg
      data-testid="leftIconButtonTestId"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M15.41 16.59 10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
      />
    </svg>
    Click
  </button>
</div>
`;

exports[`ButtonWithIcon component renders RightIcon when provided 1`] = `
<div>
  <button
    class="button button--big button--gray button--icon"
    data-testid="buttonTestId"
    type="button"
  >
    Click
    <svg
      data-testid="rightIconButtonTestId"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
      />
    </svg>
  </button>
</div>
`;

exports[`ButtonWithIcon component renders both LeftIcon and RightIcon when provided 1`] = `
<div>
  <button
    class="button button--big button--gray button--icon"
    data-testid="buttonTestId"
    type="button"
  >
    <svg
      data-testid="leftIconButtonTestId"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M15.41 16.59 10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
      />
    </svg>
    Click
    <svg
      data-testid="rightIconButtonTestId"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
      />
    </svg>
  </button>
</div>
`;

exports[`ButtonWithIcon component renders component correctly 1`] = `
<div>
  <button
    class="button button--big button--gray button--icon"
    data-testid="buttonTestId"
    type="button"
  >
    Click me
  </button>
</div>
`;
